FEATURE: Variant view in react and possibility of cropping

kitsunet opened the pull request:
This variant view based on React is a beta feature, just like the current
implementation of the variant view. It allows to crop any preset variant
with a crop, retaining the current aspect ratio.

kitsunet committed:
FEATURE: Variant view in react and possibility of cropping

This variant vew in react is beta just as intended for the variant
view. It allows to crop any preset variant with a crop, retaining
the current aspect ratio.

kitsunet committed:
Fix style errors

kitsunet committed:
Fix style errors

robertlemke commented:
@kitsunet shouldn't this be targeted to master and then downmerged into 4.3?

robertlemke commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -705,6 +717,25 @@ public function deleteAssetCollectionAction(AssetCollection $assetCollection)
         $this->forward('delete', 'AssetCollection', 'Neos.Media.Browser', ['assetCollection' => $assetCollection]);
     }
 
+    public function initializeUpdateImageVariantAction()
```
Missing Doc Block

robertlemke commented on Neos.Media.Browser/Classes/Controller/ImageVariantController.php:
```diff
@@ -0,0 +1,41 @@
+    public function updateAction(ImageVariant $imageVariant)
+    {
+        $this->assetRepository->update($imageVariant);
+        return json_encode((new ImageMapper($imageVariant))->getMappingResult());
```
Using constructor injection this way is weird, because the IDE will complain about missing arguments for ImageMapper()

kitsunet replied:
Mmm, see comment below, I am not quite happy with other options either.

robertlemke commented on Neos.Media.Browser/Classes/Domain/ImageMapper.php:
```diff
@@ -0,0 +1,100 @@
+use Neos\Media\Domain\Model\ImageVariant;
+
+/**
+ *
```
Either write something into this doc comment or leave it out

robertlemke commented on Neos.Media.Browser/Classes/Domain/ImageMapper.php:
```diff
@@ -0,0 +1,100 @@
+     * @param ResourceManager $resourceManager
+     * @param PersistenceManagerInterface $persistenceManager
+     */
+    public function __construct(ImageInterface $image, ResourceManager $resourceManager, PersistenceManagerInterface $persistenceManager)
```
Why not inject resource manager and persistence manager via setter or property injection? You could move the constructor code to `getMappingResult()` to make sure that dependencies are already injected.

kitsunet replied:
I would need to keep a reference to the image, and I didn't want to do that. Also I prefer constructor injection very much for things that are not singleton.

robertlemke commented on Neos.Media.Browser/Classes/Domain/ImageMapper.php:
```diff
@@ -0,0 +1,100 @@
+     */
+    private function mapImage(ImageInterface $image, ResourceManager $resourceManager, PersistenceManagerInterface $persistenceManager): array
+    {
+        $previeUri = $resourceManager->getPublicPersistentResourceUri($image->getResource());
```
should be `previewUri`

robertlemke reviewed (changes requested):
The feature is great and works as intended! Since this is a feature hidden behind a feature flag, I think it's okay that the UI is not fully polished. However, there are a few code style issues we should fix (see comments).

We should also fix the caption, as the dimensions information overlaps with the preset name. But I'd be fine with adjusting that in a follow-up PR.

<img width="1031" alt="image" src="https://user-images.githubusercontent.com/95582/55634622-2a350a80-57bf-11e9-911e-d4146bec5794.png">

kitsunet committed:
Fix smaller issues

kitsunet committed:
Adjust injections away from constructor injection

kitsunet committed:
Fix style issue

mficzel reviewed (changes requested):
This somehow does not work for me:

I tried these settings from the .rst documentation of the first pr https://github.com/neos/neos-development-collection/pull/2396.

Update: The docs missed the autoCreateImageVariantPresets Setting.

mficzel committed:
TASK: Improve documentation and mention the `autoCreateImageVariantPresets` option.

Also `autoCreateImageVariantPresets` is now dedined in settings but disabled by default.

mficzel reviewed (changes requested):
I generally approve and got it to work. Problem was the setting `autoCreateImageVariantPresets` that was nowhere mentioned.

The only problem left occurs when i try to update an existing crop. During the first adjustment i can select any aspect ration and only later the ratio is enforced.

mficzel commented:
@kitsunet i noticed also another weirdness. Once opening the autocreated crops not the base-image is beeing cropped but the created crop. After adjusting this is preserved.

kitsunet committed:
Adjust crop mapping to support aspect ratio

mficzel committed:
Merge branch '4.3' into feature/react-variant-view

mficzel reviewed (approved):
Works for me now and as this is behind a feature flag it can go in. Cannot say much about the js.

cxktheanna reviewed (commented):
1
